rstudio/rstudio
# bugfixes for saving/restoring window location

### Intent

Addresses [RStudio resets window location / size #12463](https://github.com/rstudio/rstudio/issues/12463)

### Approach

Ported the logic from Qt IDE for saving and restoring maximized window state.

Ported the logic from Qt IDE for determining if a window's saved position is still valid. It was WAY too aggressive; if a window extended one pixel in any direction off the display, it deemed it invalid and reset the position. Now it allows saving and restoring windows that are only partially visible on a display (used the same test as we did in Qt, namely, consider it valid if even 5 pixels are showing horizontally or vertically).

### Automated Tests

Updated unit tests.

### QA Notes

Scenarios I can think of:

Open RStudio, resize the window, close RStudio, reopen RStudio ==> should restore the size it was when closed.

Open RStudio, resize the window, maximize the window (e.g. on Microsoft Windows), close RStudio (still maximized), reopen RStudio ==> should start maximized; if you then restore the window it should be the size it was before you maximized it.

Resize/move the window so it's partially offscreen, then close and restart ==> window should restore to that same partially offscreen location.

Put the window entirely on a secondary monitor then close RStudio. Now turn off that monitor (e.g. Win+P and disable it), then restart RStudio ==> RStudio should reset position to the primary monitor, NOT display offscreen.

And so forth...

### Documentation

None

### Checklist
